docs/examples/intro_to_kernels.py

```diff
@@ -0,0 +1,328 @@
+import jax.random as jr
+from jaxtyping import install_import_hook
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import optax as ox
+from docs.examples.utils import clean_legend
+
+with install_import_hook("gpjax", "beartype.beartype"):
+    import gpjax as gpx
+
+key = jr.PRNGKey(42)
+plt.style.use(
+    "https://raw.githubusercontent.com/JaxGaussianProcesses/GPJax/main/docs/examples/gpjax.mplstyle"
+)
+cols = mpl.rcParams["axes.prop_cycle"].by_key()["color"]
+
+# %% [markdown]
+# Using Gaussian Processes (GPs) to model functions can offer several advantages over alternative methods, such as deep neural networks. One key advantage is their rich quantification of uncertainty; not only do they provide *point estimates* for the values taken by a function throughout its domain, but they provide a full predictive posterior *distribution* over the range of values the function may take. This rich quantification of uncertainty is useful in many applications, such as Bayesian optimisation, which relies on being able to make *uncertainty-aware* decisions.
+#
+# However, another advantage of GPs is the ability for one to place *priors* on the functions being modelled. For instance, one may know that the underlying function being modelled observes certain characteristics, such as being *periodic* or having a certain level of *smoothness*. The *kernel*, or *covariance function*, is the primary means through which one is able to encode such prior knowledge about the function being modelled. This enables one to equip the GP with inductive biases which enable it to learn from data more efficiently, whilst generalising to unseen data more effectively.
+#
+# In this notebook we'll develop some intuition for what kinds of priors are encoded through the use of different kernels, and how this can be useful when modelling different types of functions.
+
+# %% [markdown]
+# ## Introducing a Common Family of Kernels - The Matérn Family
+
+# %% [markdown]
+# Intuitively, the kernel defines the notion of *similarity* between the value taken at two points, $\mathbf{x}$ and $\mathbf{x}'$, by a function $f$, and will be denoted as $k(\mathbf{x}, \mathbf{x}')$:
+#
+# $$k(\mathbf{x}, \mathbf{x}') = \text{Cov}[f(\mathbf{x}), f(\mathbf{x}')]$$
+#
+#  One would expect that, given a previously unobserved test point $\mathbf{x}^*$, training points which are *closest* to this unobserved point will be most similar to it. As such, the kernel is used to define this notion of similarity within the GP framework. It tends to be up to the user to select a kernel which is appropriate for the function being modelled.
+#
+# One of the most widely used families of kernels is the Matérn family. These kernels take on the following form:
+#
+# $$k_{\nu}(\mathbf{x}, \mathbf{x'}) = \sigma^2 \frac{2^{1 - \nu}}{\Gamma(\nu)}\left(\sqrt{2\nu} \frac{|\mathbf{x} - \mathbf{x'}|}{\kappa}\right)^{\nu} K_{\nu} \left(\sqrt{2\nu} \frac{|\mathbf{x} - \mathbf{x'}|}{\kappa}\right)$$
```

Comment:
Is it worth giving the special cases where nu is half-integer valued?

Reply:
Will give an example for Matérn32